FIX for #20: Addressing silent pngquant errors

patricknelson opened the pull request:
FIX for #20: Addressing silent pngquant errors, ensuring that they don't fail silently and add option to fail grunt command if an error with pngquant is encountered.

Note that this (as it currently stands) passes the full stack trace to the console but will not result in a failure of the grunt task itself; it's annoyingly verbose, but **far** better than triggering an actual unrecoverable error for the user if they aren't prepared for it, so I find this to be a reasonable solution.

patricknelson commented:
Looking into adding a unit test for this if I can...

patricknelson commented:
Looking into adding a unit test for this if I can... just having a hard time installing a version of `node` that doesn't result in contradictory warnings/errors (due to the lack of a lock/shrinkwrap file) since one package `got@5.7.1` requires node <7 but yet `gulp-util@3.0.8` requires node >= 8 🤦‍♂️

patricknelson commented:
BTW, should this be version 1.3.1 or 1.4.0 since we're adding an option? 🤔

zauni commented:
Hey @patricknelson thanks for your contribution!

And you discovered a strange bug, because I thought if the callback gets called with an error object, it will also fail the grunt command.

I like your idea with a new option and to change the default of this option in a new major version! Sorry that there is no lock file at all, the code is not the freshest anymore :D 

And I think you are right and it should be probably version 1.4.0 because of the new option. And a working unit test would be really cool! 

In the future there could be also a Github Action which could run the unit tests on PRs. And updating all dependencies and rewriting it in TypeScript would be cool, but probably overkill for a grunt plugin. Grunt is not the cool new toy as it was 7 years ago :D

patricknelson commented:
Ok, updated docs to reference 1.4.0 since we're actually adding a feature here.

> And you discovered a strange bug, because I thought if the callback gets called with an error object, it will also fail the grunt command.

As I write my response, the song lyrics "The deeper I go..." are repeating in the song I'm listening to. Apropos... especially considering as I dig through this, I've had to rewrite my answer several times already.

Ok, final version: I've checked old versions of grunt (as far back as `grunt@0.4.3`) and even dove deep into the associated `async@0.1.22` and eventually narrowed it down to actually being a regression in this package at https://github.com/zauni/pngmin/commit/630e4d4f263f279fccaf53597a5ec39d9e2cb32f#diff-48603ff46c5bdc8ee47fb4e6c70a0435L114 (once I realized that this has nothing to do with those API's). So, starting roughly `pngmin@0.6.1` (from back on March 18, 2013). So the `callback(error)` functionality hasn't really existed since then. 

Regardless, since we're now at v1 (which lost failing errors), this means at least in some sense, we'll be remaining SemVer by defaulting this to `false`. Phew... _<wipes sweat>_

> In the future there could be also a Github Action which could run the unit tests on PRs. And updating all dependencies and rewriting it in TypeScript would be cool, but probably overkill for a grunt plugin. Grunt is not the cool new toy as it was 7 years ago :D

Yeah, I concur. Actually, `gulp` is a little more intuitive if I recall since there's a more direct/obvious path that the data takes instead of hiding things behind somewhat esoteric API's where you have to sort of "just know" how it works, it's supported a bit better via fist class language level semantics (i.e. the code vs. configuration debate). I say that especially now, considering we're talking about _JavaScript_ here, meaning you have to compensate for the lack of typing with better documentation (both of which would at least help you find where and what is going on, either automatically in the IDE or at least by documenting it).

> And I think you are right and it should be probably version 1.4.0 because of the new option. And a working unit test would be really cool!

I'll have to dig deeper into this, because this package doesn't have versioned locked down enough unfortunately for me to be able to actually run unit tests...

patricknelson commented:
Also @zauni I noticed that the indentation varies between the main task and the unit test and `Gruntfile.js` files. In the task, it's 4 spaces whereas the other JS files use 2 spaces. So... I may need to reformat some files; I think I'll opt for 4 spaces, and the reason for this is that it's much less dramatic given that it cause every single line to be modified. So instead of modifying the task file (which is much more likely to be `git blame`'d for historical reasons), I'll modify the unit test and `Gruntfile.js` instead for consistent indentation (because I'm adding `.editorconfig`).

patricknelson commented:
p.s.s. I'm also adding `yarn.lock` because I was able to get everything installed and working locally with `yarn` much more easily. Should also be a huge help moving forward.

patricknelson commented:
@zauni quick question, should these two scenarios be considered errors or should we keep these at "info" or "notice" level? Personally I don't think so, but it's right on the edge. Note that of course this would halt execution if the user set `failOnErrors: true` (so I'd prefer not right now).

```
grunt.log.writeln('No images were found in this path(s): ' + f.orig.src.join(', ').cyan);
grunt.verbose.writeln('No images were found at all...');
```

patricknelson commented:
Ok, added a unit test in 53a4c8d. Also added `yarn.lock` to lock down the dependencies and documented how to perform unit testing moving forward (using the current old/outdated `nodeunit`).

Also, interestingly this finally surfaces the underlying `pngquant` error that was occurring in `quality_test5` (so to ensure that this works as expected in v2+, I've added `failOnError: false` here just to be sure). Nice having real error reporting now for a change 😄 

![image](https://user-images.githubusercontent.com/4269377/87212968-526ebb80-c2d6-11ea-95d0-a311ca3ef5b4.png)

@zauni check this out on your system. Just clone the repo down (if you don't have it already available locally), `yarn setup`, `yarn` and then `yarn test` should get you up and running with absolutely everything (including global dependencies). LMK.

zauni commented:
Thank you very much for all the work you put into it, much appreciated!

Today I will not be able to test it out, but will come back to you as soon as I am able to.

zauni commented:
Okay, so I tested it (and funny enough, ran also into the problem with a missing libpng dependency in Ubuntu 20.04 :D) and it looks really good 👍 

Just couple of questions: It looks like the `yarn run test` command would now execute the nodeunit tests twice. Is it really necessary to have nodeunit as a global dependency and do a `grunt test && nodeunit`? Wouldn't it be enough to just use the `grunt test` command?
And I am not a yarn expert, but wouldn't yarn use the locally installed nodeunit if it is called in a package.json script? (npm would do it like that) So maybe the globally installed nodeunit dependency is not even used and the setup script could be removed.

If we have discussed that, I would merge your code and adjust the travis and appveyor CI configs to use yarn instead of npm and maybe test on more recent Node.js versions.

patricknelson commented:
> It looks like the yarn run test command would now execute the nodeunit tests twice.

Ah yes, you're _absolutely_ right!  I didn't notice it there since it looked very similar to the other tasks, but I realize it was happening two times (I'm not used to seeing unit tests rolled into a `grunt` task 😄). It's also due to my lack of understanding on how `nodeunit` works, since I didn't realize it could be run entirely locally (via that grunt task) instead of directly using `nodeunit` from the CLI (globally).

> And I am not a yarn expert, but wouldn't yarn use the locally installed nodeunit if it is called in a package.json script? (npm would do it like that) So maybe the globally installed nodeunit dependency is not even used and the setup script could be removed.

Also, yes -- removed from the `setup` script (removed that script entirely since it's best to be minimalist). At this point the only presence `yarn` has in this repository would be via the `yarn.lock` and the recommendation to use it when setting up for local development in the `README.md`. I'd personally still recommend using Yarn largely due to it's ability to lock down deep sub dependencies (particularly as this ages, it'll be a great way to ensure that you can still easily reproduce the `node_modules` for older packages as they are updated over time). Supposedly it performs better than `npm-shrinkwrap.json`; I know I've had much better success with that, working on a slightly outdated site with grunt/gulp workflows dating back to 2014. My team had constant frustration over the years, accidentally repeating work and using special hacks to get things to install. Ultimately the resolution was to move to yarn (when we could finally upgrade to a new enough version of node!).
